add landoscript

bhearsum opened the pull request:
This (fairly hefty) pull request adds a new scriptworker: landoscript. This new scriptworker will be replacing treescript when writes to hg.mozilla.org for Gecko repositories cease to happen.

Because it ultimately fulfills the same use cases as treescript, I've chosen to keep the payload as similar as possible. There are some necessary changes (mostly around using lando repository names instead of references to hg.mozilla.org repositories). I've also removed support for some things that are not used (eg: `ignore_closed_tree` and `dontbuild` are either not present for most actions, or always have the same value).

Some notes on the tests included here:
* Most of the success tests have payloads taken directly from real world treescript tasks (with some small tweaks, per the above note). I fished around and tried to find all variations of these tasks that exist (most notably for `merge_day` and `version_bump` tasks, which are often slightly different depending on which repository they are running from).
* The approach I've taken to testing here is the same as what I've been attempting to do for signingscript in https://github.com/mozilla-releng/scriptworker-scripts/pull/1097: execute tests by running `async_main`, rather than deeply unit testing individual components. This does end up making the tests notable bigger, which can be intimidating, but I've found that once written they need little to no touching, and has made refactoring as I go along significantly easier. There is probably an argument to be made that certain helper pure functions (like `diff_contents`) could do with some separate unit testing - but I've not gone to that length yet.
* I'm not terribly happy with how the responses for the github client are set-up. Because the GraphQL API uses a single endpoint for all requests, and I'm simply returning responses in a fixed order, the ordering of the `aioresponses` calls is important. In an ideal world we would be able to key responses off of parts of the requests, but I haven't had time to make this happen.

At the time of writing I have not yet been able to test it in a dev environment - I've merely been working off of the spec at https://docs.google.com/document/d/1F3xYp6v4YLLHsX7zFkYoygDyvIUPzujxeMW7W4qte0o/edit?tab=t.0#heading=h.c29tth48xps4. Seeing as this is completely unused so far, I'm hoping to get what's here merged, and then follow up with more targeted pull requests as I fix issues that I encounter when testing in the real world.

ahal commented on landoscript/src/landoscript/script.py:
```diff
@@ -0,0 +1,73 @@
+    payload = context.task["payload"]
+    artifact_dir = config["artifact_dir"]
+    public_artifact_dir = os.path.join(artifact_dir, "public", "build")
+    owner, repo = extract_github_repo_owner_and_name(payload["source_repo"])
```
I think this will need to be extracted out of the scopes instead of the payload. Or else we'll need to validate scopes independently (but then `source_repo` would need to be defined in the task definition twice)

bhearsum replied:
Ah, indeed! I have a "add scopes" todo item still; I'll make sure to address this specific point as part of that.

bhearsum replied:
I pushed an update that deals with this and the next point. I ended up adding separate `repo` and `branch` scopes (this will be needed for the main Firefox repository, where we'll have separate branches in it for various L3 repos), as well as `action` scopes. If you have any concerns please let me know!

bhearsum replied:
...actually, no `branch` scope. (After a conversation with Connor, there was a change to the Lando API: we'll be using short repo identifiers in the URLs that map to an actual github repository+branch.)

ahal commented on landoscript/src/landoscript/script.py:
```diff
@@ -0,0 +1,73 @@
+
+    lando_actions = []
+    async with GithubClient(context.config["github_config"], owner, repo) as gh_client:
+        for action in payload["actions"]:
```
Same here, I assume we'll want to protect actions via scopes too.

E.g, maybe we want to allow generating the changelog, but not bumping the version (just thinking of possible future use cases)

ahal commented on landoscript/src/landoscript/github.py:
```diff
@@ -0,0 +1,84 @@
+        if isinstance(files, str):
+            files = [files]
+
+        # Periods are not legal GraphQL key names.
```
nit: Periods and slashes

ahal commented on landoscript/src/landoscript/github.py:
```diff
@@ -0,0 +1,84 @@
+    pass
+
+
+class GithubClient:
```
Maybe it's worth putting this in `scriptworker_client.github`? The slashes fix you incorporated here is evidence that it might be

bhearsum replied:
Good call, yeah, will do.

bhearsum replied:
I moved this part out to https://github.com/mozilla-releng/scriptworker-scripts/pull/1147.

ahal commented on landoscript/src/landoscript/actions/version_bump.py:
```diff
@@ -0,0 +1,122 @@
+import logging
```
I think an argument could also be made to share the actual file manipulation logic between landoscript and treescript/githubscript. But that can be a future cleanup (and probably won't ever be a priority)

The actions you're concerned with are all Gecko specific anyway, so there's not much point. But in the future, we could have e.g a "pre-commit" action that applies lint fixes, or a "commitizen" action that generates a changelog. That kind of stuff might benefit from being shared across both scripts.

bhearsum replied:
That's a fair point. I don't really mind moving that sort of thing to shared code. It's not terribly harmful to live in `scriptworker_client` (or perhaps a new module?) even if there's never any re-use?

I know we've talked about this a bit at points, but it would probably be good to write down our long term vision for lando/github/treescript going forward. If I recall correctly, we anticipate treescript going away entirely, and any lando-enabled repo operations will move to landoscript (which will include all current supported hg repos?), and the remainder will move to githubscript? (I say this mainly selfishly, because I often get myself confused between the three of them...) The answers here may influence whether or not it's worthwhile to move some code out.

ahal replied:
Good point, could be worth writing up a quick RFC. I think you should just leave it here for now though, no need to get distracted by that yet.

bhearsum committed:
feat(landoscript): initial landoscript code with support for version bump action

This adds the rough structure for landoscript as well as implementing the `version_bump` action (necessary to make it practical to test the initial code).

bhearsum committed:
feat(landoscript): implement support for `tag` action

bhearsum committed:
refactor(landoscript): add tests that run multiple actions in one run

Most notably, this moves common set-up into conftest, and tests that aren't testing action-specific logic (eg: lando submission) into test_script.

bhearsum committed:
feat(landoscript): implement `l10n_bump` action

The helper functions here are copied out of treescript (which will soon be EOL'ed).

Also included here is some minor refactoring to avoid duplication of common of `create-commit` logic.

bhearsum committed:
feat(landoscript): implement `merge_day` action

Most of the helpers here are copied out of treescript, with some tweaks and simplications where it was possible (mostly to get rid of now-unnecessary logic).

Some refactoring of other actions was done here as well, to make it possible to call them from the `merge_day` action. Most notably: the `version_bump` action has been updated to support multiple version bumps in one run, which allows us to do all the merge day version bumps in a single commit, as we do now with treescript.

Additional test refactoring/movement was also done to make many of the helpers available to merge day tests.

bhearsum committed:
refactor(landoscript): create `run_test` helper that simplifies most of the happy path tests

This new helper allows a huge simplication for most of the action-specific tests.

bhearsum committed:
feat(landoscript): implement android l10n actions

The implementation of this is a fairly significant departure from the treescript one in two ways:
* It supports these actions without having a Gecko or `android-l10n` tree available. This necessitated fetching the remote file listings of these repositories, and using `moz.l10n` instead of `compare-locales`.
* I've fully separated out implementation of the actions. Although at a very high level they look similar, the details are different enough that IMO it's much easier and better to duplicate some of the code rather than add the indirection to avoid it. (I found it very difficult to read the treescript implementation because of this, and I didn't want to do the same here).

A necessary part of doing this was some enhancements to `diff_contents` to support added and removed files properly.

bhearsum committed:
fix(landoscript): don't support dontbuild/ignore closed tree configuration outside of l10n bump

As it turns out, these are static for all other types of actions.

bhearsum committed:
feat(landoscript): add authorization header when submitting lando requests

Also ensure that LANDO_API and LANDO_TOKEN are set up during startup.

cgsheeh commented on landoscript/src/landoscript/lando.py:
```diff
@@ -0,0 +1,77 @@
+
+
+async def submit(session, lando_api, lando_repo, actions, sleeptime_callback=calculate_sleep_time):
+    url = f"{lando_api}/api/v1/{lando_repo}"
```
We no longer use the `v1` in the URL, so this should instead look like `f"{lando_api}/api/repo/{lando_repo}"`.

cgsheeh commented on landoscript/src/landoscript/lando.py:
```diff
@@ -0,0 +1,92 @@
+                "json": json,
+                "raise_for_status": True,
+                # TODO: is this a bearer token?
+                "headers": {"Authorization": f"Bearer {lando_token}"},
```
You will need to add a `User-Agent` here that looks like `Lando-User/{username}`.

bhearsum replied:
Can I use whatever value I want here?

cgsheeh replied:
Any value is fine, but something like the email or username for the bot would be preferred. :+1:

cgsheeh commented on landoscript/src/landoscript/lando.py:
```diff
@@ -0,0 +1,92 @@
+
+        if status_resp.status == 200:
+            body = await status_resp.json()
+            if body.get("status") != "completed":
```
You should look for `"LANDED"` as the status here.

bhearsum committed:
fix(landoscript): address review comments around lando api requirements

cgsheeh commented on landoscript/src/landoscript/lando.py:
```diff
@@ -0,0 +1,95 @@
+    sleeptime_callback: Callable[..., Any] = calculate_sleep_time,
+) -> str:
+    """Submit the provided `actions` to the given `lando_repo` through the `lando_api`."""
+    url = f"{lando_api}/api/{lando_repo}"
```
This should look like `/api/repo/{lando_repo}` (unless `lando_repo` already includes `/repo`).

ahal commented on scriptworker_client/src/scriptworker_client/github_client.py:
```diff
@@ -173,3 +174,107 @@ async def get_branch_head_oid(self, branch: str) -> str:
+        )
+        file_expr = leaf_expr
+        for _ in range(depth_per_query - 1):
+            file_expr = recursive_expr.substitute(obj_expr=file_expr)
```
Maybe one day I will understand GraphQL.. but that day is not today ><.

I'm guessing you considered using https://docs.github.com/en/rest/repos/contents?apiVersion=2022-11-28 but this can get the data more efficiently?

bhearsum replied:
I did look at https://docs.github.com/en/rest/git/trees?apiVersion=2022-11-28#get-a-tree, yeah. It has a noted limit on the maximum response size (similar to the GraphQL API), and because of that I didn't bother pursuing it further, as we most likely would've needed to recursively fetch through it as well, and I thought it was nicer to avoid needing an HTTP and GraphQL client.

bhearsum replied:
(also, I should've mentioned this upfront: this part of the change is in its own pull request over at https://github.com/mozilla-releng/scriptworker-scripts/pull/1161)

bhearsum replied:
I do recall now doing some experiments with the HTTP API as well, and finding that it may have also been slower, in addition to being just as complex...although that is a weak memory at this point.

ahal replied:
> I thought it was nicer to avoid needing an HTTP and GraphQL client.

Fyi the simple-github client supports both. IIRC they even share an aiohttp session. Not that that changes anything here :)

ahal commented on scriptworker_client/src/scriptworker_client/github_client.py:
```diff
@@ -173,3 +174,107 @@ async def get_branch_head_oid(self, branch: str) -> str:
 
         return repo["object"]["oid"]
+
+    async def get_file_listing(self, path: str = "", branch: Optional[str] = None, depth_per_query=10) -> List[str]:
```
Please add a docstring similar to the other functions

bhearsum replied:
I addressed this and the other comments about this commit in https://github.com/mozilla-releng/scriptworker-scripts/pull/1161.

ahal commented on scriptworker_client/src/scriptworker_client/github_client.py:
```diff
@@ -173,3 +174,107 @@ async def get_branch_head_oid(self, branch: str) -> str:
+        return files
+
+
+def process_file_listings(entries: List[Dict[str, Any]], prefix: Path = Path("")) -> Tuple[List[str], List[Path]]:
```
nit: it seems a bit strange for this to be a module function. Seems like it should be a private method on the client class.

bhearsum replied:
Yeah, that's fair. At the very least it needs to be prefixed; but I don't see any reason it can't go on the class.

cgsheeh commented on landoscript/docker.d/worker.yml:
```diff
@@ -0,0 +1,12 @@
+  autoland:
+    owner: mozilla-firefox
+    repo: firefox
+    branch: autoland
```
The current output from this endpoint (the URL is `/api/repoinfo/<phab_short_name>`) looks like this:

```
{
	"repo_url": "https://github.com/mozilla-firefox/firefox",
	"branch_name": "main",
	"scm_level": "main.scm_level_3"
}
```

I guess the `owner` and `repo` fields you have here are derived from the `repo_url` we have. Should I split those out on our side so they don't need to be parsed from the URL?

I can also remove the `main.` part from the `scm_level`, that's a detail of our permissions implementation.

bhearsum replied:
Thank you! Feel free to structure this however you want; I don't mind needing to do a bit of parsing on our side as long the information is there, and unambigious (as it is now).

On a related note: is there going to be a `central` phab short name? Or is `mozilla-central` going to go by `main` in the new world?

cgsheeh replied:
Okay, I removed the `main.` part from the scm level, it will now return a string like `"scm_level_3"`. I'll keep the `repo_url` field the same and let you handle parsing the `owner` and `repo` values out of that field. :)

> On a related note: is there going to be a central phab short name? Or is mozilla-central going to go by main in the new world?

`mozilla-central` will go by `main` in the new world, and the others repos will use names that match their bookmarks in mozilla unified (ie `mozilla-beta` is the `beta` branch, `mozilla-esr128` is the `esr128` branch).

bhearsum replied:
Cheers!

bhearsum replied:
I'm getting rid of the `central` references in https://github.com/mozilla-releng/scriptworker-scripts/pull/1164.

ahal commented on landoscript/src/landoscript/actions/version_bump.py:
```diff
@@ -0,0 +1,143 @@
+
+def log_file_contents(contents):
+    for line in contents.splitlines():
+        log.info(line)
```
It's not clear to me why this is better than just logging contents directly. Having the logger preamble in front of each line just makes it harder to copy/paste imo.

Feel free to ignore this if this was just copy/pasted and you want to keep it as similar as possible.

bhearsum replied:
I don't recall now...my only goal is to make sure newlines are rendered. That can probably be done without the splitting though!

bhearsum replied:
Oh, I recall why we have this now: it's to ensure that we get one log line per line in the file, while ensuring each line is prefixed. (Arguably we could simply log with a leading newline to achieve a similar effect...)

bhearsum replied:
Just for clarity: I plan to leave this as is unless you feel it's preferable to log with a preceding newline.

ahal commented on landoscript/src/landoscript/lando.py:
```diff
@@ -0,0 +1,76 @@
+
+
+async def poll_until_complete(session: ClientSession, poll_time: int, status_url: str):
+    while True:
```
So I guess the idea is that we poll until we hit the Taskcluster timeout?

I think we might be losing some of the idempotent-ness of  the old approach here. It will certainly be possible for this task to fail due to timeout, but Lando could still land it afterwards. This means a failed task doesn't necessarily indicate a failed landing.

Do you know what would happen if we re-ran the task in such a case? For version bump, it might be worth ensuring that if the requested version is already what's committed, that we exit successfully. I'm not really sure what to do in the general case..

Another approach might be to timeout within the task, and abort the landing prior to failing. That way there's more of a guarantee that a failed landoscript task means the push didn't get landed. I don't think this needs to be fully solved for the MVP

bhearsum replied:
Yeah, polling until we timeout is what is happening here, and that's a really good point you've made.

And to be honest, I'm not actually sure what the best thing to do here is. I think in the short term, attempting to cancel the request at shutdown for any non-success makes sense. But this is also less than ideal if, eg: we get spot terminated and have to redo the work because of it. There's also no guarantees that we'll be able to cancel the request in time in the event of a spot termination.

Another obvious option (for the future) would be to take the notarization approach, and do the polling in a separate task. But I also wonder if we even need to bother to poll in some cases? For example, l10n bumps don't have anything downstream of them...maybe we could just spit out the status URL, and leave it at that for cases like that?

jcristau replied:
Spot termination is not a concern for scriptworkers.  We might get killed if we exceed terminationGracePeriodSeconds though, in which case we should still aim to preserve idempotence.  It looks like we only ever have a single lando request, which hopefully is handled atomically at the other end (TIL git push has a `--atomic` option, nice, I'm only 9 years late to notice)?  (I was wondering how this worked for merge day.  Currently we have 2 hg pushes, e.g. once to tag central and another to merge to beta.  I guess that goes away because we know they're the same underlying git repo and so it doesn't matter where the tag goes?)

We could try and cancel the submission on shutdown (bitrisescript seems to do something similar: https://github.com/mozilla-releng/scriptworker-scripts/blob/9867329bf5b2956a90bd990ded424473c42fc2f2/bitrisescript/src/bitrisescript/script.py#L47), and there should be a couple of minutes between SIGTERM and SIGKILL.  I would be a bit worried about races between the attempt to cancel and the request landing (or maybe the cancel attempt itself times out, or ...).

So there's a few scenarios that could happen AFAICT:
1. maybe the initial request didn't go through, we can rerun everything, all good
2. maybe the initial request failed, or was canceled, we can safely retry.
3. maybe the initial request succeeded before the rerun, we should detect that and not do anything
4. maybe the initial request is still in flight.  We could potentially find it and poll on it, or cancel it and retry?  If we can't find it, we may have 2 requests in flight, hopefully only one can succeed?

In treescript we also had the case where one push went through but not the other, and the case where when the next run checked version numbers, the first push had gone through to hgssh but was not yet mirrored out to hgweb.

bhearsum replied:
Thanks for the additional thoughts and strategies, @jcristau! Not needing to worry about spot termination certainly simplifies things, as we can make some safer assumptions.

I'm going to give this a bit more thought - but clearly we need some sort of special handling for the rerun scenario, if not separating polling altogether.

> In treescript we also had the case where one push went through but not the other, and the case where when the next run checked version numbers, the first push had gone through to hgssh but was not yet mirrored out to hgweb.

If we're talking about two pushes to the same repository, this case will go away AFAIK - all changes to a given repository (which in the new world means branch within the new github repo) will either happen or not.

bhearsum replied:
>  Currently we have 2 hg pushes, e.g. once to tag central and another to merge to beta. I guess that goes away because we know they're the same underlying git repo and so it doesn't matter where the tag goes?)

Yeah, that's right. Although that does bring to mind something a bit odd about the API: `tag` actions are applied to a repository+branch - not to a repository overall. @cgsheeh - is this something you've thought about? (I don't think we necessarily need to change anything here, I'm just noting it because it's a bit odd.)

bhearsum replied:
The more I think about how to handle termination, the more it seems like moving polling out of the landoscript tasks might be the best thing to do. Something like:
* Return immediately from landoscript after a successful submission; attaching the status url in a well enough artifact
* Add a new task downstream of each landoscript task which does the polling. This doesn't need to be a scriptworker task AFAIK, because polling doesn't require credentials.

It would also avoid the need for any cancellation in landoscript, and thus any race conditions associated with that. If we trust that a landoscript task would never be rerun unless a human has inspected the state of things first, this would eliminate the need for special rerun handling in landoscript, I think. (Although perhaps it's too much a leap to assume it would never be errantly rerun, and we'd want at least some of the rerun safety checks?)

The main downside here is that if the polling task finds that the landoscript submission has failed, we'd need to rerun the upstream landoscript task, and then force a new polling task afterwards.

bhearsum replied:
For posterity, I discovered after writing this that all Lando endpoints (including status polling) require authentication. So at least in the immediate term, moving polling outside of landoscript is not an option. (I guess we could poll in a separate landoscript task, but that probably doesn't provide any benefits...)

ahal commented on landoscript/src/landoscript/script.py:
```diff
@@ -0,0 +1,104 @@
+# `context` is kept explicitly untyped because all of its members are typed as
+# Optional. This never happens in reality (only in tests), but as things stand
+# at the time of writing, it means we need noisy and unnecessary None checking
+# to avoid linter complaints.
```
Might be worth a test fixture to provide dummy values for everything in context? I don't really care that much though, if it's super annoying don't bother.

bhearsum replied:
I think changing this would be a larger effort around improving types in scriptworker (the problematic class being referenced here is https://github.com/mozilla-releng/scriptworker/blob/bba07b47036bfa4870741ba54c6d5e98e91134f3/src/scriptworker/context.py#L57-L64, and I haven't looked recently, but I think there's assumptions in tests over there as well around certain things being nullable that don't really make sense in reality).

ahal commented on landoscript/src/landoscript/script.py:
```diff
@@ -0,0 +1,104 @@
+
+    # Note: `lando_repo` is not necessarily the same as the repository's name
+    # on Github.
+    lando_repo = payload["lando_repo"]
```
It's a common pattern to just parse this out of the scopes in other scriptworkers, that way you don't need to validate it independently, nor have it defined twice in the task definition.

I'm a bit on the fence on which way is better, so if you prefer this approach, I'm fine with leaving it.

bhearsum replied:
If we want to do it for consistency I could be convinced, but my personal view here is that it's weird and surprising that scopes shouldn't affect behaviour. (Sorry...I should've called this out upfront, as it is different than what we do in at least some other scriptworkers.)

ahal commented on landoscript/src/landoscript/script.py:
```diff
@@ -0,0 +1,104 @@
+        for action in payload["actions"]:
+            log.info(f"processing action: {action}")
+
+            if action == "version_bump":
```
I think it'll be cleaner to create some kind of module level mapping to call these, rather than if/else chaining.

bhearsum replied:
I think this is a good idea...it will require the `run` functions to all have the same interface though. This is easy enough when it comes to the return values, but as things stand now the inputs vary. (Obviously I could just pass the entire `config` rather than the specific bits needed, but that harms readability in the `run` functions IMO.)

bhearsum replied:
(Oh, you touched on this in a later comment...whoops!)

ahal commented on landoscript/src/landoscript/script.py:
```diff
@@ -76,6 +76,9 @@ async def async_main(context):
                     lando_actions.append(version_bump_action)
+            elif action == "tag":
+                tag_actions = tag.run(payload["tags"])
+                lando_actions.extend(tag_actions)
```
Going back to the module level mapping, I don't think it would hurt to generalize the `None` test to all actions here (even if the action implementation would never return `None`)

ahal commented on landoscript/src/landoscript/actions/l10n_bump.py:
```diff
@@ -0,0 +1,231 @@
+    l10n_repo_url: str
+    l10n_repo_target_branch: str
+    ignore_config: dict[str, list[str]]
+    platform_configs: list[PlatformConfig]
```
Tbh I'm not really sure what `TypedDict` is meant for, but these look like they should be dataclasses.

ahal commented on landoscript/src/landoscript/script.py:
```diff
@@ -64,42 +67,58 @@ async def async_main(context):
+                    )
+                    # sometimes nothing has changed!
+                    if l10n_bump_actions:
+                        lando_actions.extend(l10n_bump_actions)
```
I guess the actions are different enough that my previous suggestion of generalizing this might be a bit of a pain?

I still think it would be do-able, and that there's probably some benefit to having a standardized interface around calling actions and handling their results.. But I also don't want to cause a major refactor of this stack if things are already working.

So I guess feel free to leave it for now :)

ahal commented on landoscript/src/landoscript/actions/merge_day.py:
```diff
@@ -0,0 +1,285 @@
+    touch_clobber_file: bool
+    version_files: list[VersionFile]
+    replacements: list[list[str]]
+    regex_replacements: list[list[str]]
```
Same comment, these should likely be dataclasses.

ahal commented on landoscript/src/landoscript/actions/android_l10n_import.py:
```diff
@@ -0,0 +1,106 @@
+
+class AndroidL10nImportInfo(TypedDict):
+    from_repo_url: str
+    toml_info: list[TomlInfo]
```
These look like dataclasses.

ahal commented on landoscript/src/landoscript/actions/android_l10n_import.py:
```diff
@@ -0,0 +1,106 @@
+            log.info(f"processing toml file: {toml_file}")
+
+            if toml_contents[toml_file] is None:
+                raise LandoscriptError(f"toml_file '{toml_file}' is not present in repository")
```
The variable repetition is a bit odd here. Couldn't you just iterate over `toml_contents.items()` instead of accessing `android_l10n_import_info` multiple times and then needing validation?

bhearsum replied:
For checking for None here? Yeah, you're right, that would be better. (We still need this loop to pull `dest_path` for each toml file though, of course, but I don't think you were suggesting otherwise?)

bhearsum replied:
This is getting handled in https://github.com/mozilla-releng/scriptworker-scripts/pull/1165

ahal commented on landoscript/src/landoscript/actions/android_l10n_import.py:
```diff
@@ -0,0 +1,106 @@
+from pathlib import Path
+from typing import TypedDict
+
+import tomli
```
IIRC we're now using Python 3.11 for all scriptworkers other than iscript right?

If that's the case, just use the stdlib:
https://docs.python.org/3/library/tomllib.html

bhearsum replied:
Ah, excellent!

bhearsum replied:
Handling this in https://github.com/mozilla-releng/scriptworker-scripts/pull/1166

ahal commented on landoscript/src/landoscript/actions/android_l10n_sync.py:
```diff
@@ -0,0 +1,95 @@
+class AndroidL10nSyncInfo(TypedDict):
+    from_repo_url: str
+    from_branch: str
+    toml_info: list[TomlInfo]
```
Ok, so I finally looked it up. I guess one reason to use `TypedDict` is if, well, you need a dict rather than a class. I'm not really sure why that would be necessary, but maybe avoiding converting everything is reason enough.

It's worth noting that you can convert a dataclass to a dict with the `dataclasses.asdict` module function.

bhearsum replied:
Yeah, there's no reason these have to be TypedDicts - I've converted them. (It turns out TypedDict handles nesting a bit more gracefully, but dataclasses deal with required/optional/defaults much better IMO, so it's still preferable.)

ahal commented on landoscript/src/landoscript/actions/android_l10n_sync.py:
```diff
@@ -0,0 +1,95 @@
+        log.info(f"processing toml file: {toml_file}")
+
+        if toml_contents[toml_file] is None:
+            raise LandoscriptError(f"toml_file '{toml_file}' is not present in repository")
```
Same comment about the odd variable shuffling as before.

bhearsum replied:
This is getting handled in https://github.com/mozilla-releng/scriptworker-scripts/pull/1165

ahal commented on landoscript/src/landoscript/actions/android_l10n_sync.py:
```diff
@@ -0,0 +1,95 @@
+    diff = ""
+    for l10n_file in l10n_files:
+        if l10n_file["dst_name"] not in orig_files:
+            log.warning(f"WEIRD: {l10n_file['dst_name']} not in dst_files, continuing anyways...")
```
Lol, I approve of this new log level.

ahal commented on landoscript/tests/conftest.py:
```diff
@@ -126,26 +126,51 @@ async def run_test(
+        type_ = "blob" if len(parts) == 1 else "tree"
+        obj = {}
+        if type_ == "tree":
+            # this obviously does not handle multiple files in the same directory
```
I'm not sure "obviously" applies here ><

ahal commented on landoscript/docker.d/init_worker.sh:
```diff
@@ -1,2 +1,20 @@
+  fi
+}
+
+# TODO: real URLs
```
Comment just to help remember this.

ahal reviewed (changes requested):
Looking good! Did a first pass

bhearsum committed:
feat(landoscript): pull repository url and branch from lando instead of config

bhearsum committed:
fix(landoscript): Replace TypedDicts with dataclasses

This required a little bit of massaging for places where there's nested dataclasses, but it was otherwise straightforward.

This change also uncovered some places where tests were including unnecessary data, which have been fixed. It also replaced some existing null checks (because the dataclass will throw if a required member is not present during construction).

bhearsum committed:
fix(landoscript): improve comment around some test code

ahal reviewed (approved):
Thanks, lgtm!
